### ISSUE TITLE
FROST super issue

### ISSUE THREAD

**Issue by LLFourn** [LLFourn/secp256kfun#85]
<a id="org7156b95"></a>

## secp256kfun FROST impl

<a id="org0490987"></a>

### TODO Notion DKG id

This is useful so we can sign it with pop and in gun to make sure every device displays the same thing before using they key.

<a id="org557c3f0"></a>

### TODO Proof of possession

Just use hash type parameter that is passed to internal Schnorr instance. Tag the hash when you init Frost with &ldquo;frost/dkg&rdquo;. Then hash all the commitments together to form the dkg-id, use the internal Schnorr (with noncegen) to produce pop.

```rust
\#[derive(Clone, Debug, Default)]
pub struct Frost<H, NG = ()> {
    schnorr: Schnorr<H, NG>,
    dkg_id_hash: H
}
```

<a id="org0f46ea5"></a>

### TODO How do we make it two rounds in gun

-   Problem is we need to negate our own first coef based on the joint key which we don&rsquo;t know yet.

<a id="orge89027c"></a>

### TODO Have API that allows you to pass in first coef

Need to handle needs negation later:

    // X = (b*x) * G where b is 1 or -1.
    // Then you tweak
    // X' = X + t * G
    // if X' needs negation then sk is
    // -(b*x + t) = -b*x - t
    // new b = -b i.e. prev needs_negation xor new needs_negation.
    // Store new join_public_key and new tweak, as well as needs_negation

Stop using needs negation when creating shares.

<a id="org33e06ac"></a>

### TODO How should gen<sub>nonce</sub> be designed

    pub fn gen_nonce(
            &self,
            secret: &Scalar,
            sid: &[u8]
        ) -> NonceKeyPair { ... }

Application makes sure sid is unique. Document `secret` could be either static first coef or secret
share depending on the application &#x2013; but be consistent!

<a id="org436517d"></a>

### TODO Renaming things

-   s/JointKey/FrostKey/
-   s/Dkg/KeyGen/
-   s/collect<sub>polys</sub>/new<sub>keygen</sub>/
-   s/FirstRoundError/NewKeyGenError/
-   s/collect<sub>shares</sub>/finish<sub>keygen</sub>/
-   s/SecondRoundError/FinishKeyGenError/

<a id="org0293a38"></a>

### Don&rsquo;t use HashMap

Doesn&rsquo;t work with no<sub>std</sub> and introduces non-determinism. Probably just use vec internally and warn users.

<a id="orgf8d25d3"></a>

### Nonce agg scheme what to hash?

No

<a id="orgb6a3e87"></a>

### Do proptesting

-   Random thresholds up to some degree e.g. 100
-   Random number of participants
-   Test independent for correct error when participants less than threshold
-   Random tweaks &#x2013; 0..=2 of them

**Comment by nickfarrow** [comment]
Also:
### Self Proving & Validation
Currently user creates and validates their own shares + pop. Is this annoying?

### Tweak Tests

### Added a secret.clone() when creating the proptest
bad?

**Comment by nickfarrow** [comment]
All of these have been implemented in https://github.com/LLFourn/secp256kfun/pull/84 except for additional tweak tests (no test vectors yet)

**Comment by LLFourn** [comment]
closed by #84. More things to be done but they are not captured by this issue.
